src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,418 @@
+idea is that the user of the library should not be given a way
+to produce a cyclic graph.
+
+We define a 'Num' instance as a convenient notation for working
+with graphs:
+
+  > edgeList   0           == []
+  > vertexList 0           == [0]
+  > edgeList   (1 + 2)     == []
+  > vertexList (1 + 2)     == [1,2]
+  > edgeList   (1 * 2)     == [(1,2)]
+  > vertexList (1 * 2)     == [1,2]
+  > edgeList   (1 + 2 * 3) == [(2,3)]
+  > vertexList (1 + 2 * 3) == [1,2,3]
+  > edgeList   (1 * 2 + 3) == [(1,2)]
+  > vertexList (1 * 2 + 3) == [1,2,3]
+
+The default partial order for the 'Num' instance to restrict
+edges is (\<), ie. Edges can only be formed from vertex
+numbered /a/ to a vertex numbered /b/ if and only if 
+/a/ \< /b/:
+
+  > edgeList (1 * 2) == [(1,2)]
+  > edgeList (2 * 1) == []
+  > (2 * 1) == (2 + 1)
+
+__Note:__ the 'Num' instance does not satisfy several "customary laws"
+of 'Num', which dictate that 'fromInteger' @0@ and 'fromInteger' @1@
+should act as additive and multiplicative identities, and 'negate' as
+additive inverse. Nevertheless, overloading 'fromInteger', '+' and '*'
+is very convenient when working with algebraic graphs; we hope that
+in future Haskell's Prelude will provide a more fine-grained class
+hierarchy for algebraic structures, which we would be able to utilise
+without violating any laws.
+-}
+newtype AdjacencyMap a = AAM
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show, Eq, Ord)
+
+-- | Check if the internal graph representation is consistent,
+-- i.e. that all edges refer to existing vertices and the graph
+-- is acyclic. It should be impossible to create an inconsistent 
+-- adjacency map.
+--
+-- @
+-- consistent 'empty'           == True
+-- consistent (1 + 2)         == True
+-- consistent (1 * 2 + 2 * 3) == True
+-- @
+consistent :: Ord a => AdjacencyMap a -> Bool
+consistent (AAM m) = AM.consistent m && AM.isAcyclic m
+
+-- | Construct the /empty acyclic graph/.
+-- Complexity: /O(1)/ time and memory.
+empty :: AdjacencyMap a
+empty = coerce AM.empty
+
+-- | Construct the graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+vertex :: a -> AdjacencyMap a
+vertex = coerce AM.vertex
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is
+-- the length of the given list.
+--
+-- @
+-- vertices []        == 'empty'
+-- vertices [x]       == 'vertex' x
+-- vertices [1, 2, 3] == 1 + 2 + 3
+-- @
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices = coerce AM.vertices
+
+-- | Perform a disjoint overlay of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edgeList (overlayD empty empty)             == []
+-- edgeList (overlayD (1 * 2 + 1 * 3) (1 * 2)) == [(Left 1,Left 2) 
+--                                                ,(Left 1,Left 3)
+--                                                ,(Right 1,Right 2)]
+-- @
+overlayD ::
```

Comment:
I had a look at some standard libraries, and it seems more common to use prefix "disjoint" rather than suffix "D". So, let's rename these functions to `disjointOverlay` and `disjointConnect`.

Reply:
Actually, an even better option would be to use "union" and "join" instead, which are standard terms from graph theory:

http://mathworld.wolfram.com/GraphUnion.html
http://mathworld.wolfram.com/GraphJoin.html

What do you think?

Reply:
Since we are distinguishing between which `AdjacencyMap` the vertex came from, using `disjointUnion` makes more sense. `union` and `join` are similar to `overlay` and `connect`.

Reply:
To me `disjointUnion` feels redundant, because graph union is disjoint by definition. I guess we better stick to `disjointOverlay` to avoid confusing terms. It's fairly easy to understand the difference between `overlay` and `disjointOverlay`.